### PR TITLE
Genai intro

### PR DESCRIPTION
**Pull request by meken**

### CODE REVIEW

File: hacks/genai-intro/README.md [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+### Tips
+
+- Check the provided Cloud Function configuration to see the details on how it's being triggered
+
+## Challenge 2: First steps into the LLM realm
+
+### Introduction
+
+Let's get started with a simple objective; we're going to _extract_ the title of a document using LLMs. In order to work with LLMs we need textual data, so the first step in our process is to extract text data from PDF documents. We've already implemented that functionality for you using Cloud Vision APIs in the provided Cloud Function. Go ahead and have a look at the `extract_text_from_document` to understand where and how the results are stored. Now, with those results we can look into extracting the title from the text content of the document.
```
Comment (gfilicetti) [inline hacks/genai-intro/README.md:65]: "... look at the extract_text_from_document function to understand..."

File: hacks/genai-intro/README.md [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+The _Refine_ chain approach also makes multiple calls to an LLM, but it does that in an iterative fashion. It starts with the first document/chunk, passes its content and gets a response, and then gets to the second document/chunk passing that content plus the response from the previous call, iterating until the last document/chunk and then passing the last (rolling) response and getting a final answer using a different (final) prompt.
+
+### Description
+
+In order to get the summaries, we'll implement the _Refine_ approach for this challenge. Most of the code is already provided in the `extract_summary_from_text` method in Cloud Function. Similar to the previous challenge, you're expected to design the prompts and provide the right mapping.
+
+### Success Criteria
+
+- For this [paper](https://arxiv.org/pdf/2310.01473) we expect a summary like this:
```
Comment (gfilicetti) [inline hacks/genai-intro/README.md:108]: Add an empty line between 108 and the text block. It'll render better.

File: hacks/genai-intro/README.md [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+  | Full-Atom Protein Pocket Design via Iterative Refinement | Quantitative Biology |
+
+### Learning Resources
+
+- Creating BigQuery [datasets](https://cloud.google.com/bigquery/docs/datasets) and [tables](https://cloud.google.com/bigquery/docs/tables)
+- BigQuery [LLM support](https://cloud.google.com/bigquery/docs/generate-text-tutorial)
+
+### Tips
+
+- You could download and upload the papers manually, but you can also consider `wget` and `gsutil` from Cloud Shell.
```
Comment (gfilicetti) [inline hacks/genai-intro/README.md:173]: "... also consider using wget ..."

File: hacks/genai-intro/solutions.md [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    """
+```
+
+And make sure to truncate the text:
+
+```python
+response = model.predict(prompt.format(text=text[:10000]))
+```
+
+If you want to use gsutil & jq to get the contents, see:
```
Comment (gfilicetti) [inline hacks/genai-intro/solutions.md:54]: "... get the contents, this is the command to use:"

File: hacks/genai-intro/solutions.md [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        summaries += f"\n{context}"
+    
+    prompt = final_prompt_template.format(text=summaries)
+    return model.predict(prompt).text   
+```
+
+## Challenge 4: BigQuery &#10084; LLMs
+
+### Notes & Guidance
+
```
Comment (gfilicetti) [inline hacks/genai-intro/solutions.md:111]: Also add some lead in text here, whatever you think is appropriate, I don't have a good example

**Review by gfilicetti — COMMENTED** [review]
Just minor stuff

File: hacks/genai-intro/README.md [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+  The author argues that the standard cosmological model is incorrect and that there is no dark matter. The author provides several arguments for this, including:
+
+  * The observed properties of galaxies are consistent with them being self-regulated, largely isolated structures that sometimes interact.
+  * The observed uniformity of the galaxy population is evidence against the standard cosmological model.
+  * The large observed ratio of star-forming galaxies over elliptical galaxies is evidence against the standard cosmological model.
+
+  The author concludes that understanding galaxies purely as baryonic, self-gravitating systems becomes simple and predictive.
+  ```
+  > **Note** By their nature, LLM results can vary, this is something to expect so your exact text may not match the above, but the intent should be the same.
+
```
[the remark this answers is not in the document]
Comment (meken) [reply hacks/genai-intro/README.md:119]: There is, but all of that point to langchain, and I don't want to complicate things/confuse people by pointing to some documentation that uses a library we don't include.

**Pull request merged**